mmaction/datasets/samplers/distributed_sampler.py

```diff
@@ -40,43 +43,92 @@ def __iter__(self):
-    from the entire dataset.
+    The default value of dynamic_length is True, which means we use
+    oversampling / subsampling, and the dataset length may changed. If
+    dynamic_length is set as False, the dataset length is fixed.
     """
 
-    def __init__(self, dataset, num_replicas=None, rank=None, power=1, seed=0):
+    def __init__(self,
+                 dataset,
+                 num_replicas=None,
+                 rank=None,
+                 dynamic_length=True,
+                 shuffle=True,
+                 seed=0):
         super().__init__(dataset, num_replicas=num_replicas, rank=rank)
-        self.power = power
+        self.shuffle = shuffle
+
+        if type(dataset).__name__ == 'RepeatDataset':
+            dataset = dataset.dataset
+
+        assert hasattr(dataset, 'class_prob')
+
+        self.class_prob = dataset.class_prob
+        self.dynamic_length = dynamic_length
+        # for the compatibility from PyTorch 1.3+
         self.seed = seed if seed is not None else 0
 
     def __iter__(self):
-        # deterministically shuffle based on epoch
         g = torch.Generator()
-        g.manual_seed(self.epoch + self.seed)
-        video_infos_by_class = self.dataset.video_infos_by_class
-        num_classes = self.dataset.num_classes
-        # For simplicity, discontinuous labels are not permitted
-        assert set(video_infos_by_class) == set(range(num_classes))
-        counts = [len(video_infos_by_class[i]) for i in range(num_classes)]
-        counts = [cnt**self.power for cnt in counts]
-
-        indices = torch.multinomial(
-            torch.Tensor(counts),
-            self.total_size,
-            replacement=True,
-            generator=g)
-        indices = indices.data.numpy().tolist()
+        g.manual_seed(self.seed + self.epoch)
+
+        class_indices = defaultdict(list)
+
+        # To be compatible with RepeatDataset
+        times = 1
+        dataset = self.dataset
+        if type(dataset).__name__ == 'RepeatDataset':
+            times = dataset.times
+            dataset = dataset.dataset
+        for i, item in enumerate(dataset.video_infos):
+            class_indices[item['label']].append(i)
+
+        if self.dynamic_length:
+            indices = []
+            for k, prob in self.class_prob.items():
+                prob = prob * times
+                for i in range(int(prob // 1)):
+                    indices.extend(class_indices[k])
+                rem = int((prob % 1) * len(class_indices[k]))
+                rem_indices = torch.randperm(
+                    len(class_indices[k]), generator=g).tolist()[:rem]
+                indices.extend(rem_indices)
+            if self.shuffle:
+                shuffle = torch.randperm(len(indices), generator=g).tolist()
+                indices = [indices[i] for i in shuffle]
+
+            # re-calc num_samples & total_size
+            self.num_samples = math.ceil(len(indices) / self.num_replicas)
```

Comment:
Can num_replicas be None?

Reply:
It cannot be None, since it is only called in build_dataloader, and we specify these args there.